LDAP state module improvements

rhansen opened the pull request:
### What does this PR do?

Various improvements to the `ldap` state module and `ldap3` execution module:
  * Redo and expand tests.
  * Documentation improvements.
  * Code health improvements.
  * Improvements to attribute value ordering in `ldap.managed`.
  * Improvements to encoding/decoding of attribute values.
  * Work around a flaw in the way `python-ldap` generates modification lists.

There are now integration tests, which requires OpenLDAP to be installed.  The tests use a custom Docker image to get OpenLDAP (see https://github.com/saltstack/salt-ci-containers/pull/3).

### What issues does this PR fix or reference?
None.

### Merge requirements satisfied?
**[NOTICE] Bug fixes or features added to Salt require tests.**
<!-- Please review the [test documentation](https://docs.saltproject.io/en/master/topics/tutorials/writing_tests.html) for details on how to implement tests into Salt's test suite. -->
- [x] Docs
- [x] Changelog - https://docs.saltproject.io/en/master/topics/development/changelog.html
- [x] Tests written/updated

### Commits signed with GPG?
No

Please review [Salt's Contributing Guide](https://docs.saltproject.io/en/master/topics/development/contributing.html) for best practices.

See GitHub's [page on GPG signing](https://help.github.com/articles/signing-commits-using-gpg/) for more information about signing commits with GPG.

rhansen committed:
ldap: Ignore order when checking equality of attribute value sets

rhansen committed:
ldap: Minor documentation improvements

  * Clarify that the `_format_unicode_password()` return value is
    double quoted, not the argument.
  * Refine the CLI examples:
      * Wrap lines with line continuations for readability.
      * Add missing backslashes before line continuations.
      * Add missing quotes.
      * Add missing square brackets for list values.
  * Rewrap some comments at column 80.
  * Miscellaneous wording and formatting refinements.

rhansen committed:
ldap: Replace `OrderedDict` with `OrderedSet` where appropriate

Now that an `OrderedSet` class exists, use it instead of `OrderedDict`
in the places where `OrderedDict` was used to emulate an ordered set.

rhansen committed:
ldap: Move `LDAPError` class to a utility module

rhansen committed:
ldap: Fix `str` <-> `bytes` encoding/decoding

Move all encoding and decoding logic to the `AttributeValueSet` class
and consistently use it when reading from or writing to LDAP.  This
fixes some encoding/decoding corner cases, and improves the API's
usability.

Technically this is a backwards-incompatible change: `ldap3.search`
and `ldap.managed` now return decoded strings when possible.  However,
it appears that the Salt master (or maybe the master/minion protocol?)
automatically decodes returned `bytes` objects to `str` when possible,
so this change only affects direct function calls.

rhansen committed:
ldap: Redo `change()` to avoid flaw in `python-ldap` package

rhansen committed:
ldap: Use standard `OrderedDict` class

rhansen committed:
ldap: Convert unit tests to integration tests

The existing unit tests I wrote long ago were overly complicated and
not that great, so this commit converts them into integration tests
against a real OpenLDAP server.  Docker is required to run the
integration tests; the tests are skipped if Docker isn't available.

rhansen committed:
ldap: Add integration tests for `ldap3.search`

MKLeb commented:
Hi @rhansen, thanks for your contribution!

It looks like there are some failing tests here.  And for the integration tests, perhaps use a docker container for openldap?  Something like https://hub.docker.com/r/bitnami/openldap/ should work, and there are some examples around the test suite where we use pytest-salt-factories to spin up docker containers in our test suite!

MKLeb commented:
Here is an example of using saltfactories for the docker containers - https://github.com/saltstack/salt/blob/18e878dba65fcfc83adb2e3614fe111003b3ec0e/tests/pytests/functional/cache/test_consul.py#L41

rhansen commented:
Thanks for the pointer @MKLeb.

I'm having a bit of trouble figuring out how to set up tests.  The `python-ldap` package must be installed on a machine with OpenLDAP installed because it compiles some C code against an OpenLDAP library.  I think the most straightforward way to handle this is to create a custom Docker container that is both a minion and an OpenLDAP server, and control the minion from a test master daemon.  Any tips on how to do that?  (e.g., pointers to existing test code that does something similar)  Or do you have an alternative suggestion?

MKLeb commented:
> Thanks for the pointer @MKLeb.
> 
> I'm having a bit of trouble figuring out how to set up tests. The `python-ldap` package must be installed on a machine with OpenLDAP installed because it compiles some C code against an OpenLDAP library. I think the most straightforward way to handle this is to create a custom Docker container that is both a minion and an OpenLDAP server, and control the minion from a test master daemon. Any tips on how to do that? (e.g., pointers to existing test code that does something similar) Or do you have an alternative suggestion?

Apologies, I should have looked a bit further here. It seems OpenLDAP is already installed into the CI machines (see [here](https://github.com/saltstack/salt-jenkins/blob/d5c35ad9948829e706e1d7785ec52e21ca6327a6/state-tree/golden-images-provision.sls#L61)).

From what I can tell, your integration tests are being skipped because it can't find the ldapadd command.  Perhaps we can use the ldap3 module itself (and its add method) to add the entries instead of ldapadd from slapdtest?  It all depends what underlying tools it uses to do that.  I get that it's kind of meta to use the ldap3 mod to test it, but I think that's our best and quickest route here.

rhansen commented:
> It seems OpenLDAP is already installed into the CI machines (see [here](https://github.com/saltstack/salt-jenkins/blob/d5c35ad9948829e706e1d7785ec52e21ca6327a6/state-tree/golden-images-provision.sls#L61)).

It looks like `libldap2-dev` (on Debian-based systems anyway) needs to be installed as well:

```
  gcc -pthread -Wno-unused-result -Wsign-compare -DNDEBUG -g -fwrapv -O3 -Wall -fPIC -DHAVE_SASL -DHAVE_TLS -DLDAPMODULE_VERSION=3.4.3 "-DLDAPMODULE_AUTHOR=python-ldap project" "-DLDAPMODULE_LICENSE=Python style" -IModules -I/__w/salt/salt/.nox/docs-html-compress-false-clean-true/include -I/usr/local/include/python3.8 -c Modules/LDAPObject.c -o build/temp.linux-x86_64-cpython-38/Modules/LDAPObject.o
  In file included from Modules/LDAPObject.c:3:
  Modules/common.h:15:10: fatal error: lber.h: No such file or directory
   #include <lber.h>
            ^~~~~~~~
  compilation terminated.
  error: command '/usr/bin/gcc' failed with exit code 1
  ----------------------------------------
  ERROR: Failed building wheel for python-ldap
```

It's probably easy to add that package to the golden images, but I think it would be nicer if the tests tested against a Docker container rather than requiring OpenLDAP to be installed on the CI machines.  Getting that to work will take some time and effort; I'll start working on that.  If you'd prefer, we can do the easy thing first (install `libldap2-dev` on the CI machines) and migrate the tests to Docker in a future pull request.

MKLeb commented:
I agree, docker would be the nice way to do it, and likely the correct way as well.  There is an example of how we run a minion in a docker container [here](https://github.com/saltstack/salt/blob/a1a07d1c8c07d5df9ff30c74ba5d75530f019a86/tests/pytests/scenarios/compat/test_with_versions.py#L62).  It uses an image built in https://github.com/saltstack/salt-ci-containers.  We can't do quite what those do in that repo though, because we likely want to test against a minion running the changes on the current branch, as is the case here for instance.  You can probably just pip install the local salt, all the ldap dependencies, and pip install python-ldap.  I would imagine you would either create an inline dockerfile and build it each time or create a minimal dockerfile that runs the minion on the local salt and then use that minion to then install the ldap dependencies and python-ldap.  Not sure which is better honestly, but for no reason at all I'm leaning towards the second option :)

[Here](https://pytest-salt-factories.readthedocs.io/en/latest/ref/saltfactories/daemons/container.html) is the docs on how the container minion works.  Let me know if you need any help!

rhansen committed:
ldap: Preserve the order of multi-valued attributes

rhansen commented:
OK, I think I got it working.  It depends on https://github.com/saltstack/salt-ci-containers/pull/3 so the tests will fail until that is merged and an image is published.

rhansen commented:
I'll rebase onto latest master and fix some of the lint errors.

rhansen commented:
Looks like the new `openldap-minion-*` container is leaking into the `tests.pytests.integration.netapi.*` tests.  The `openldap_minion` fixture has scope `package`, so it should be shut down by the time those tests run.  Any ideas?

MKLeb commented:
> Looks like the new `openldap-minion-*` container is leaking into the `tests.pytests.integration.netapi.*` tests. The `openldap_minion` fixture has scope `package`, so it should be shut down by the time those tests run. Any ideas?

Hey @rhansen, thanks for diving deep into this!  From a cursory glance, I'm not sure what is going on there.  Maybe I can dig around here tomorrow and look for a fix.  I'll let you know as soon as I do.

MKLeb commented:
Aha!  I got it to work.  I switched them to module scopes, and also added a line to delete the minion's key after we are done with it (that is what was messing up the netapi tests).  Here is my diff, and you'll notice I switched from using salt_cilent to salt_cli just to be a little more explicit and consisten across the test suite.

```
diff --git a/tests/pytests/integration/modules/test_ldap3.py b/tests/pytests/integration/modules/test_ldap3.py
index e45d6617be..b591636dc5 100644
--- a/tests/pytests/integration/modules/test_ldap3.py
+++ b/tests/pytests/integration/modules/test_ldap3.py
@@ -11,7 +11,7 @@ pytestmark = [
 
 
 def test_search(openldap_minion_run, subtree, u0dn):
-    assert openldap_minion_run("ldap3.search", base=subtree) == {
+    assert openldap_minion_run("ldap3.search", base=subtree).data == {
         subtree: {
             "objectClass": ["dcObject", "organization"],
             "dc": ["test_search"],
@@ -29,7 +29,7 @@ def test_search(openldap_minion_run, subtree, u0dn):
 def test_search_filter(openldap_minion_run, subtree, u0dn):
     assert openldap_minion_run(
         "ldap3.search", base=subtree, filterstr="(sn=Lastname)"
-    ) == {
+    ).data == {
         u0dn: {
             "objectClass": ["person"],
             "cn": ["u0"],
diff --git a/tests/support/pytest/ldap.py b/tests/support/pytest/ldap.py
index 6ec208bb79..2d470b1c3c 100644
--- a/tests/support/pytest/ldap.py
+++ b/tests/support/pytest/ldap.py
@@ -202,8 +202,8 @@ class SlapdMinion(SaltMinion):
         self.__run_unpriv(cmd)
 
 
-@pytest.fixture(scope="package")
-def openldap_minion(salt_master, salt_client):
+@pytest.fixture(scope="module")
+def openldap_minion(salt_master, salt_cli, salt_key_cli):
     name = random_string("openldap-minion-")
     c = salt_master.salt_minion_daemon(
         name,
@@ -218,15 +218,15 @@ def openldap_minion(salt_master, salt_client):
     # keyword argument 'python_executable'".
     c.python_executable = "python3"
     with c.started():
-        ret = salt_client.run("test.ping", minion_tgt=c.id)
-        assert ret is True
+        ret = salt_cli.run("test.ping", minion_tgt=c.id)
+        assert ret.data is True
         yield c
+    salt_key_cli.run("-y", "-d", c.id)
 
-
-@pytest.fixture(scope="package")
-def openldap_minion_run(openldap_minion, salt_client):
+@pytest.fixture(scope="module")
+def openldap_minion_run(openldap_minion, salt_cli):
     def _run(fn, *args, **kwargs):
-        return salt_client.run(
+        return salt_cli.run(
             fn,
             *args,
             minion_tgt=openldap_minion.id,
@@ -237,8 +237,8 @@ def openldap_minion_run(openldap_minion, salt_client):
     yield _run
 
 
-@pytest.fixture(scope="package")
-def openldap_minion_apply(openldap_minion_run, salt_client):
+@pytest.fixture(scope="module")
+def openldap_minion_apply(openldap_minion_run):
     def _apply(fn, **kwargs):
         has_name = "name" in kwargs
         kwargs.setdefault("name", "x")

```

rhansen commented:
> Aha! I got it to work.

Awesome, thank you!  I rebased and pushed the fixes; hopefully the tests will all pass.

> I switched them to module scopes,

Oops, I conflated "package" with "module" again.  I meant to do module, but thought it was spelled "package".  :facepalm: 

> and also added a line to delete the minion's key after we are done with it (that is what was messing up the netapi tests).

Ah, makes sense!  I had expected the master to be cleaned up after module scope (due to confusing package with module) so I didn't think I needed to clean up the key.

> I switched from using salt_cilent to salt_cli just to be a little more explicit and consistent across the test suite.

I intentionally used `salt_client` because tuples can't be serialized to JSON, and I wanted to pass a non-list object to make sure that it can accept arbitrary iterables.  Perhaps it would be better to check that in a unit test instead?

Thanks for the help!

MKLeb commented:
> I intentionally used `salt_client` because tuples can't be serialized to JSON, and I wanted to pass a non-list object to make sure that it can accept arbitrary iterables. I added a comment to make that clear, though perhaps it would be better to do the non-list iterable check in a unit test instead?
>

Oops, I missed that comment.  If you wanted to make that a unit test that should be fine!  Also, is there a reason you deleted the old ldap module unit test file?  If they were incompatible with your changes, could you add a few more simple ones to make sure we have some coverage there as well?  Thanks!  Let me know if you need any more help.

rhansen commented:
> > I intentionally used `salt_client` because tuples can't be serialized to JSON, and I wanted to pass a non-list object to make sure that it can accept arbitrary iterables. I added a comment to make that clear, though perhaps it would be better to do the non-list iterable check in a unit test instead?
> 
> Oops, I missed that comment.

That's because I added the comment in response to your suggestion to use `salt_cli` instead of `salt_client`. :smile: 

> If you wanted to make that a unit test that should be fine!

I added some unit tests for `salt.states.ldap._toset()`.  Now that there is test coverage for various input types, I will change the integration tests to use `salt_cli` so that they're more integration-y and less unit-y.

> Also, is there a reason you deleted the old ldap module unit test file?

They're redundant with the new integration tests, and they're not very readable/maintainable.  I wasn't happy with them when I first wrote them several years ago, but I kept them anyway because they were better than nothing.  Now they don't add anything, so there's no real point in keeping them.

Well, they do one thing better: they're lighter weight, so they're OK for smoke testing.

> If they were incompatible with your changes,

No, just redundant.

> could you add a few more simple ones to make sure we have some coverage there as well?

I'll add some lighter tests—tests that run without needing to spin up a Docker container or pass `--run-slow --run-destructive`.

ITJamie commented:
out of interest. will this help with the following. when setup master ldap servers and adding certain config it currently can throw issues with attempting to manage a var that already exists (see commented out field).

it would also be nice to not have to specifically declare the object id (since depending on the order the item was added the id might be different). maybe by regex? eg `olcOverlay={.}syncprov,olcDatabase={1}mdb,cn=config` instead of `olcOverlay={0}syncprov,olcDatabase={1}mdb,cn=config`
```
manage_ldap_database_syncprov_overlay_create:
  ldap.managed:
    - connect_spec:
        url: ldapi:///
        bind:
          method: sasl
          mechanism: EXTERNAL
    - entries:
        - 'olcOverlay={0}syncprov,olcDatabase={1}mdb,cn=config':
            - delete_others: False
            - default:
                objectClass:
                  - olcOverlayConfig
                  - olcSyncProvConfig
                olcOverlay: syncprov
                olcSpCheckpoint: "100 10"
                # if ever setting up a new master. this will need to be uncommented. then re-commented once it exists.
                # olcSpSessionLog: 100
 ```

rhansen commented:
@ITJamie I don't quite understand the problem you're describing.  What does the `olcSpSessionLog` attribute have to do with the numbering in the entry's DN?

rhansen committed:
ldap: Fix support for bare `bytes` values, add test coverage

rhansen committed:
ldap: Exercise non-list iterable in integration test

rhansen committed:
ldap: Define a new class for holding attribute values

Future commits will modify this class to fix some bugs.

Ch3LL commented:
Is this ready for review? I noticed its still in draft status?

rhansen committed:
yaml: Convert `salt.utils.yaml` tests to pytest

rhansen commented:
> Is this ready for review? I noticed its still in draft status?

Not yet.  There are still some tweaks I want to make to the tests.  I'll unmark as draft and @ mention once it's ready for review.

rhansen committed:
yaml: Fix custom YAML to object constructor registration

The `SaltYamlSafeLoader.add_constructor()` method is a class method,
not an instance method.  Therefore, any call to that method in an
instance method (such as `__init__`) affects all future instances, not
just the current instance (`self`).  That's not a problem if the
registration calls simply re-register the same constructor functions
over and over, but that was not the case here: different functions
were registered depending on the value of the `dictclass` parameter to
`__init__`.

The net effect was that an `!!omap` node was correctly processed as a
sequence of single-entry maps (and a list of (key, value) tuples
returned) until the first time a SaltYamlSafeLoader was constructed
with a non-`dict` class.  After that, every `!!omap` node was always
incorrectly processed as a map node regardless of `dictclass`.

Now `!!omap` processing is performed as originally intended:
  * If the `dictclass` parameter is `dict` (the default), the behavior
    when loading a `!!map` or `!!omap` node is unchanged from the base
    class's behavior.
  * If the `dictclass` parameter is not `dict`:
      * `!!map` nodes are loaded like they are in the base class
        except the custom class is used instead of `dict`.
      * `!!omap` nodes are loaded the same as `!!map` nodes.  (This is
        a bug because an `!!omap` node is a sequence node of
        single-valued map nodes, not an ordinary map node.  A future
        commit will fix this.)

Behavior before:

```
$ python
>>> import salt.utils.yaml as y
>>> import collections
>>> y.load("!!omap [{foo: bar}, {baz: bif}]")
[('foo', 'bar'), ('baz', 'bif')]
>>> y.SaltYamlSafeLoader("", dictclass=collections.OrderedDict)  # created for side-effect only
<salt.utils.yamlloader.SaltYamlSafeLoader object at 0xc31a10>
>>> y.load("!!omap [{foo: bar}, {baz: bif}]")  # exact same as before
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "salt/utils/yamlloader.py", line 159, in load
    return yaml.load(stream, Loader=Loader)
  File "venv/lib/python3.8/site-packages/yaml/__init__.py", line 81, in load
    return loader.get_single_data()
  File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 51, in get_single_data
    return self.construct_document(node)
  File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 60, in construct_document
    for dummy in generator:
  File "salt/utils/yamlloader.py", line 45, in construct_yaml_map
    value = self.construct_mapping(node)
  File "salt/utils/yamlloader.py", line 56, in construct_mapping
    raise ConstructorError(
yaml.constructor.ConstructorError: expected a mapping node, but found sequence
  in "<unicode string>", line 1, column 1
```

Behavior after:

```
$ python
>>> import salt.utils.yaml as y
>>> import collections
>>> y.load("!!omap [{foo: bar}, {baz: bif}]")
[('foo', 'bar'), ('baz', 'bif')]
>>> y.SaltYamlSafeLoader("!!omap [{foo: bar}, {baz: bif}]", dictclass=collections.OrderedDict).get_single_data()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 51, in get_single_data
    return self.construct_document(node)
  File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 60, in construct_document
    for dummy in generator:
  File "salt/utils/yamlloader.py", line 42, in construct_yaml_omap
    return (yield from self.construct_yaml_map(node))
  File "salt/utils/yamlloader.py", line 36, in construct_yaml_map
    value = self.construct_mapping(node)
  File "salt/utils/yamlloader.py", line 52, in construct_mapping
    raise ConstructorError(
yaml.constructor.ConstructorError: expected a mapping node, but found sequence
  in "<unicode string>", line 1, column 1
>>> y.load("!!omap [{foo: bar}, {baz: bif}]")
[('foo', 'bar'), ('baz', 'bif')]
```

This commit also adds a unit test for the `dictclass` parameter,
though it's important to note that the new test is not a regression
test for the bug fixed by this commit.  (I don't think it would be
worthwhile to write a regression test because the test code would be
complicated and unreadable.)

rhansen committed:
yaml: Load `!!omap` nodes as sequences of mappings

Before, `!!omap` nodes were incorrectly assumed to be mapping nodes
when `dictclass` was not `dict`.  Now they are correctly processed as
sequences of single-entry mappings.  The resulting Python object has
type `dictclass` (usually `collections.OrderedDict` or a subclass).

This commit does not change the behavior when `dictclass` is `dict`;
loading an `!!omap` node still returns a list of (key, value)
tuples (PyYAML's default behavior).  I consider that to be a bug in
PyYAML, so a future commit may change the behavior in the `dict` case
to match the non-`dict` behavior.  (This commit uses `dictclass` for
the non-`dict` case to match what appears to be the original
intention.)

Behavior before:

```
$ python
>>> import salt.utils.yaml as y
>>> import collections
>>> y.load("!!omap [{foo: bar}, {baz: bif}]")
[('foo', 'bar'), ('baz', 'bif')]
>>> y.SaltYamlSafeLoader("!!omap [{foo: bar}, {baz: bif}]", dictclass=collections.OrderedDict).get_single_data()
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 51, in get_single_data
    return self.construct_document(node)
  File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 60, in construct_document
    for dummy in generator:
  File "salt/utils/yamlloader.py", line 42, in construct_yaml_omap
    return (yield from self.construct_yaml_map(node))
  File "salt/utils/yamlloader.py", line 36, in construct_yaml_map
    value = self.construct_mapping(node)
  File "salt/utils/yamlloader.py", line 52, in construct_mapping
    raise ConstructorError(
yaml.constructor.ConstructorError: expected a mapping node, but found sequence
  in "<unicode string>", line 1, column 1
```

Behavior after:

```
$ python
>>> import salt.utils.yaml as y
>>> import collections
>>> y.load("!!omap [{foo: bar}, {baz: bif}]")
[('foo', 'bar'), ('baz', 'bif')]
>>> y.SaltYamlSafeLoader("!!omap [{foo: bar}, {baz: bif}]", dictclass=collections.OrderedDict).get_single_data()
OrderedDict([('foo', 'bar'), ('baz', 'bif')])
```

Relevant bug: https://github.com/saltstack/salt/issues/12161

rhansen committed:
yaml: Load `!!omap` nodes as `collections.OrderedDict` objects

Now the behavior is consistent regardless of the value of the
Loader's `dictclass` constructor parameter.

Starting with Python 3.6, Python `dict` objects always iterate in
insertion order, so iteration order was already guaranteed when using
an ordinary YAML map.  However, `!!omap` provides a stronger guarantee
to users (plain map iteration order can be thought of as a Salt
implementation detail that might change in the future), and it allows
them to make their `.sls` files self-documenting.  For example,
instead of:

```yaml
my_pillar_data:
  key1: val1
  key2: val2
```

users can now do:

```yaml
my_pillar_data: !!omap
  - key1: val1
  - key2: val2
```

to make it clear to readers that the entries are intended to be
processed in order.

Behavior before:

```
$ python
>>> import salt.utils.yaml as y
>>> import collections
>>> y.load("!!omap [{foo: bar}, {baz: bif}]")
[('foo', 'bar'), ('baz', 'bif')]
>>> y.SaltYamlSafeLoader("!!omap [{foo: bar}, {baz: bif}]", dictclass=collections.OrderedDict).get_single_data()
OrderedDict([('foo', 'bar'), ('baz', 'bif')])
```

Behavior after:

```
$ python
>>> import salt.utils.yaml as y
>>> import collections
>>> y.load("!!omap [{foo: bar}, {baz: bif}]")
OrderedDict([('foo', 'bar'), ('baz', 'bif')])
>>> y.SaltYamlSafeLoader("!!omap [{foo: bar}, {baz: bif}]", dictclass=collections.OrderedDict).get_single_data()
OrderedDict([('foo', 'bar'), ('baz', 'bif')])
```

Relevant bug: https://github.com/saltstack/salt/issues/12161

rhansen committed:
yaml: Add integration test for YAML map iteration order

This demonstrates that https://github.com/saltstack/salt/issues/12161
has already been fixed (thanks to Python 3.6 changing `dict` to
iterate in insertion order).

rhansen committed:
yaml: Fix IndentedSafeOrderedDumper indentation

Before, the indentation would only be increased if the compiled module
`yaml.CSafeDumper` did not exist.  Now it is indented even if it does.

Behavior before (assuming `yaml.CSafeDumper` exists):

```
$ python
>>> import salt.utils.yaml as y
>>> print(y.dump({"foo": ["bar"]}, Dumper=y.IndentedSafeOrderedDumper, default_flow_style=False))
foo:
- bar

```

Behavior after:

```
$ python
>>> import salt.utils.yaml as y
>>> print(y.dump({"foo": ["bar"]}, Dumper=y.IndentedSafeOrderedDumper, default_flow_style=False))
foo:
  - bar

```

rhansen committed:
yaml: Use a `for` loop to factor out duplicate code

rhansen committed:
yaml: Dump `OrderedDict` objects as `!!omap` nodes

Before, an `OrderedDict` object was represented as a plain YAML
mapping.  Now it is represented as an `!!omap` node, which is a
sequence of single-valued mappings.

This is a behavior change: Any users that depend on an `OrderedDict`
becoming a plain YAML mapping (which would be read in as a `dict`
object) must first convert the `OrderedDict` to `dict`.

This change preserves the semantics of the object, and it preserves
round-trip identity.

Behavior before:

```
$ python
>>> from salt.utils.odict import OrderedDict
>>> import salt.utils.yaml as y
>>> import collections
>>> print(y.dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
foo: bar

>>> print(y.safe_dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
foo: bar

>>> print(y.dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
foo: bar

>>> print(y.safe_dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
foo: bar

```

Behavior after:

```
$ python
>>> from salt.utils.odict import OrderedDict
>>> import salt.utils.yaml as y
>>> import collections
>>> print(y.dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
!!omap
- foo: bar

>>> print(y.safe_dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
!!omap
- foo: bar

>>> print(y.dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
!!omap
- foo: bar

>>> print(y.safe_dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
!!omap
- foo: bar

```

rhansen committed:
yaml: Dump all `OrderedDict` types the same way

Behavior before:

```
$ python
>>> from salt.utils.odict import OrderedDict
>>> import salt.utils.yaml as y
>>> import collections
>>> print(y.dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
foo: bar

>>> print(y.safe_dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
foo: bar

>>> print(y.dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
!!python/object/apply:collections.OrderedDict
- - - foo
    - bar

>>> print(y.safe_dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
NULL

```

Behavior after:

```
$ python
>>> from salt.utils.odict import OrderedDict
>>> import salt.utils.yaml as y
>>> import collections
>>> print(y.dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
foo: bar

>>> print(y.safe_dump(OrderedDict([("foo", "bar")]), default_flow_style=False))
foo: bar

>>> print(y.dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
foo: bar

>>> print(y.safe_dump(collections.OrderedDict([("foo", "bar")]), default_flow_style=False))
foo: bar

```

rhansen committed:
yaml: Load `!!python/tuple` nodes as `tuple` objects

My main motivation for adding this is to facilitate testing, though it
also gives module authors greater flexibility (in particular: a
`tuple` can be a `dict` key, but a `list` cannot because it is not
hashable), and I don't see a strong reason why this shouldn't be
added.

Behavior before:

```
$ python
>>> import salt.utils.yaml as y
>>> y.load("!!python/tuple [foo, bar]")
Traceback (most recent call last):
  File "<stdin>", line 1, in <module>
  File "salt/utils/yamlloader.py", line 159, in load
    return yaml.load(stream, Loader=Loader)
  File "venv/lib/python3.8/site-packages/yaml/__init__.py", line 81, in load
    return loader.get_single_data()
  File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 51, in get_single_data
    return self.construct_document(node)
  File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 55, in construct_document
    data = self.construct_object(node)
  File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 100, in construct_object
    data = constructor(self, node)
  File "venv/lib/python3.8/site-packages/yaml/constructor.py", line 427, in construct_undefined
    raise ConstructorError(None, None,
yaml.constructor.ConstructorError: could not determine a constructor for the tag 'tag:yaml.org,2002:python/tuple'
  in "<unicode string>", line 1, column 1
  ```

Behavior after:

```
$ python
>>> import salt.utils.yaml as y
>>> y.load("!!python/tuple [foo, bar]")
('foo', 'bar')
```

rhansen committed:
yaml: Default to `OrderedDumper` in `salt.utils.yaml.dump()`

I believe this was the original intention.  Even if it was not, the
symmetry with `salt.utils.yaml.save_dump()` makes `dump()` less
surprising now, and it makes it easier to introduce representer
changes that affect both `dump()` and `safe_dump()`.

rhansen committed:
yaml: Factor out duplicate code in `salt.utils.yaml.safe_dump()`

rhansen committed:
yaml: Improve readability of `salt.utils.yaml.dump()`

rhansen committed:
yaml: Delete unnecessary `IndentMixin` class to improve readability

rhansen committed:
yaml: Add TODO comments next to puzzling code

rhansen committed:
yaml: Load `!!timestamp` nodes as `datetime.datetime` objects

YAML nodes tagged with `!!timestamp` nodes now become Python
`datetime` objects rather than strings.  To preserve compatibility
with existing YAML files, timestamp-like strings without `!!timestamp`
are still loaded as strings.

Behavior before:

```
$ python
>>> import datetime
>>> import salt.utils.yaml as y
>>> y.load("'2022-10-21T18:16:03.1-04:00'")
'2022-10-21T18:16:03.1-04:00'
>>> y.load("!!timestamp '2022-10-21T18:16:03.1-04:00'")
'2022-10-21T18:16:03.1-04:00'
```

Behavior after:

```
$ python
>>> import datetime
>>> import salt.utils.yaml as y
>>> y.load("'2022-10-21T18:16:03.1-04:00'")
'2022-10-21T18:16:03.1-04:00'
>>> y.load("!!timestamp '2022-10-21T18:16:03.1-04:00'")
datetime.datetime(2022, 10, 21, 18, 16, 3, 100000, tzinfo=datetime.timezone(datetime.timedelta(days=-1, seconds=72000)))
```

rhansen committed:
yaml: Dump `datetime.datetime` objects with `!!timestamp` tag

Delete the ineffectual representer, and unregister the implicit
resolver for timestamp strings so that `datetime.datetime` objects are
always written with an accompaying `!!timestamp` tag.

This is a behavior change: Any users that depend on a `datetime`
becoming a plain YAML string (which would be read in as a `str`
object) must convert the object to string themselves.

This change preserves the semantics of the object, and it preserves
round-trip identity.

Behavior before:

```
$ python
>>> import datetime
>>> import salt.utils.yaml as y
>>> print(y.dump(datetime.datetime(2022, 10, 21, 18, 16, 3, 100000, tzinfo=datetime.timezone(datetime.timedelta(hours=-4)))))
2022-10-21 18:16:03.100000-04:00

```

Behavior after:

```
$ python
>>> import datetime
>>> import salt.utils.yaml as y
>>> print(y.dump(datetime.datetime(2022, 10, 21, 18, 16, 3, 100000, tzinfo=datetime.timezone(datetime.timedelta(hours=-4)))))
!!timestamp 2022-10-21 18:16:03.100000-04:00

```

rhansen committed:
yaml: Dump `tuple` objects as `!!python/tuple` nodes

Before, `!!python/tuple` was only used for `OrderedDumper`.  Now it is
also used for `SafeOrderedDumper` and `IndentedSafeOrderedDumper`.

This is a behavior change: Any users that depended on a `tuple`
becoming a plain YAML sequence (which would be read in as a `list`
object) must first convert the `tuple` to `list`.

This change preserves the semantics of the object, and it preserves
round-trip identity.  Preserving round-trip identity is particularly
important if the tuple is used as a key in a `dict` because `list`
objects are not hashable.

Behavior before:

```
$ python
>>> import salt.utils.yaml as y
>>> print(y.dump(("foo", "bar"), default_flow_style=False))
!!python/tuple
- foo
- bar

>>> print(y.safe_dump(("foo", "bar"), default_flow_style=False))
- foo
- bar

```

Behavior after:

```
$ python
>>> import salt.utils.yaml as y
>>> print(y.dump(("foo", "bar"), default_flow_style=False))
!!python/tuple
- foo
- bar

>>> print(y.safe_dump(("foo", "bar"), default_flow_style=False))
!!python/tuple
- foo
- bar

```

rhansen committed:
yaml: Fix documentation about `datetime` conversion

I believe the behavior changed with commit
002aa88a97ed8a1c51f4ab1c965d22d064ea30a8 which was first released in
Salt v2018.3.0.

rhansen committed:
yaml: Document that `!!omap` should be avoided due to bugs

MKLeb commented:
@rhansen Are there any updates here?

rhansen commented:
Not yet—I got distracted by some yak shaving:  https://github.com/saltstack/salt/pull/62932

rhansen commented:
@MKLeb For some reason Python `logging` log messages in the containerized minion are not showing up in the test log, which makes troubleshooting difficult.  The plumbing that sets up pytest logging is quite complicated and magical, so I'm having trouble locating the problem.  My best guess is that [`saltfactories.utils.saltext.log_handlers.pytest_log_handler.setup_handlers()`](https://github.com/saltstack/pytest-salt-factories/blob/1.0.0rc20/src/saltfactories/utils/saltext/log_handlers/pytest_log_handler.py#L78) is supposed to be called but is not.  Do you have any suggestions?

rhansen commented:
Figured it out:  https://github.com/saltstack/pytest-salt-factories/issues/139

I still want to add some more unit tests, and https://github.com/saltstack/salt/pull/62932 needs to be merged.  Once those are done, this should be ready to go.

MKLeb commented:
Sorry for not getting back to you quickly, but glad to hear you figured it out.

> Figured it out: [saltstack/pytest-salt-factories#139](https://github.com/saltstack/pytest-salt-factories/issues/139)
> 
> I still want to add some more unit tests, and #62932 needs to be merged. Once those are done, this should be ready to go.

Going to look at https://github.com/saltstack/salt/pull/62932 today and suggest changes (if any).  We don't need to solve that salt factories issue first, correct?

rhansen commented:
> We don't need to solve that salt factories issue first, correct?

Correct. I have a workaround in place.

s0undt3ch commented on tests/support/pytest/ldap.py:
```diff
@@ -0,0 +1,392 @@
+                # Bind mount the checked-out source code into the Docker
+                # container so that a minion daemon started inside the container
+                # will run the code to be tested.
+                RUNTIME_VARS.CODE_DIR: {
```
Don't use `RUNTIME_VARS`, that's something we inherited from the old test suite.
```python
from tests.conftest import CODE_DIR
```

s0undt3ch commented on tests/support/pytest/ldap.py:
```diff
@@ -0,0 +1,392 @@
+        v = importlib_metadata.version("pytest-salt-factories")
+        for cmd in [
+            ("install_packages", "python3-pip"),
+            ("pip", "install", f"pytest-salt-factories=={v}"),
```
Salt factories is bind mounted on the minion container

https://github.com/saltstack/pytest-salt-factories/blob/1.0.0rc20/src/saltfactories/daemons/container.py#L772

It's not importable but log forwarding ["should"](https://github.com/saltstack/pytest-salt-factories/blob/1.0.0rc20/src/saltfactories/manager.py#L221-L225) still [work](https://github.com/saltstack/pytest-salt-factories/blob/1.0.0rc20/src/saltfactories/manager.py#L237-L243).

s0undt3ch commented on tests/pytests/integration/modules/test_ldap3.py:
```diff
@@ -0,0 +1,39 @@
+import pytest
+
+pytest_plugins = [
+    "tests.support.pytest.ldap",
+]
```
```suggestion
from tests.support.pytest.ldap import openldap_minion_run
```
Import the fixture instead.
